perceiver/data/audio/symbolic.py

```diff
@@ -0,0 +1,216 @@
+        return encode_midi_files(midi_files, num_workers=self.hparams.preproc_workers)
+
+    def _to_flattened_numpy_array(self, input_ids_list: List[np.ndarray]) -> np.ndarray:
+        input_ids = [np.append(input_ids, [self._EXAMPLE_SEPARATOR_INPUT_ID]) for input_ids in input_ids_list]
+        return np.concatenate(input_ids)
+
+    @staticmethod
+    def _save_memory_map(data: np.ndarray, target_file: Path) -> None:
+        fp = np.memmap(str(target_file.absolute()), dtype=np.int16, mode="w+", shape=data.shape)
+        fp[:] = data[:]
+        fp.flush()
+
+    def setup(self, stage: str = None) -> None:
+        self._ds_train = SymbolicAudioNumpyDataset(
+            data_file=str(self.train_data_file),
+            max_seq_len=self.hparams.max_seq_len + 1,
+            separator_input_id=self._EXAMPLE_SEPARATOR_INPUT_ID,
+            min_seq_len=self.hparams.min_seq_len + 1 if self.hparams.min_seq_len is not None else None,
+        )
+        self._ds_valid = SymbolicAudioNumpyDataset(
+            data_file=str(self.valid_data_file),
+            max_seq_len=self.hparams.max_seq_len + 1,
+            separator_input_id=self._EXAMPLE_SEPARATOR_INPUT_ID,
+        )
+
+    def train_dataloader(self):
+        return DataLoader(
+            self._ds_train,
+            shuffle=False,
+            collate_fn=self._collator,
+            batch_size=self.hparams.batch_size,
+            num_workers=self.hparams.num_workers,
+            pin_memory=self.hparams.pin_memory,
+        )
+
+    def val_dataloader(self):
+        return DataLoader(
+            self._ds_valid,
+            shuffle=False,
+            collate_fn=self._collator,
+            batch_size=self.hparams.batch_size,
+            num_workers=self.hparams.num_workers,
+            pin_memory=self.hparams.pin_memory,
+        )
+
+
+class SymbolicAudioNumpyDataset(torch.utils.data.Dataset):
+    def __init__(self, data_file: str, max_seq_len: int, separator_input_id: int, min_seq_len: Optional[int] = None):
+        self._data = np.memmap(data_file, dtype=np.int16, mode="r")
+        self._max_seq_len = max_seq_len
+        self._separator_input_id = separator_input_id
+        self._min_seq_len = min_seq_len
+        self._length = self._data.shape[0] // self._max_seq_len
+
+    def __getitem__(self, index):
+        start_pos = torch.randint(self._data.shape[0] - self._max_seq_len, (1,)).item()
+        sample = torch.tensor((self._data[start_pos : start_pos + self._max_seq_len]).astype(np.int64))
+
+        if self._separator_input_id not in sample:
+            example = sample
+        else:
+            examples = list(
+                torch.tensor_split(sample, torch.where(sample == self._separator_input_id)[0])
+            )  # split at example separators
+            examples = sorted(examples, key=lambda x: len(x), reverse=True)  # order by longest sample
+
+            example = examples[0]
+            example = example[example != self._separator_input_id]
```

Comment:
Why is this needed? Doesn't `example != self._separator_input_id` always evaluate to a tensor filled with to `True` here?

Reply:
The `sample` is split into individual `examples` at each occurrence of the `self._separator_input_id` using `torch.tensor_split`. `torch.tensor_split` preserves the occurrences of each `self._separator_id` which is why they are removed here.

Reply:
`self._separator_input_id` doesn't occur in the first split element, only in elements 2..n. This is why you can omit this line. Or am I missing something?

Reply:
Hmmm ... do I miss the special case when the separator is the very first element in the entire sequence? Can this happen at all?

Reply:
The split elements are sorted by length, so the first split element after the sort operation might contain the separator_id.

Reply:
:+1: Ah right, this is what I was missing ...